hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -330,13 +396,56 @@ private void initialize(ClientProxy clientProxy) throws Exception {
     }
 
     public Collection<? extends DistributedObject> getDistributedObjects() {
+        return getLocalDistributedObjects();
+    }
+
+    private void getDistributedObjectsAndCleanStaleLocalProxies() {
+        ClientMessage request = ClientGetDistributedObjectsCodec.encodeRequest();
+        distributedObjectsWriteLock.lock();
```

Comment:
You are holding a write lock even if the invocation would return proxy list which will not cause any modification to your local list and you will do this periodically which will block all lock read requests (a.g. client.getMap("myMap).get("myKey);) periodically.

Reply:
In the latest commit, only proxy creations will block. If a proxy exists, it won't deal with any lock mechanism and return the proxy.

Reply:
That is what I am saying, just get the write lock and block all proxy creations only if the proxies obtained from remote is different from the local ones.

Reply:
When we discussed with @sancar, he said that we should lock before sending a request. 

Indeed, I think it fails in the following scenario:

1.lets say we lock after getting response, no matter whether it is conditional or not.

2. while the request is in flight client creates some proxies.

3. server responds to getDistObjects call, new proxies not created yet

4. server creates proxies.

5.  client gets dist objects and process them. since new proxy is not in the response, it is removed from local proxies.

6. if it is a Cache proxy it will throw error once it is used(Sancar's revert is due to Cache proxy) Cache proxies need local proxy to exist